### ISSUE TITLE
Fix pyflakes for core/python

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6988]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#6988|
|Type|task|
|Created|2014-02-27 11:52:22Z|
|Branch|https://github.com/twisted/twisted/tree/python-pyflakes-6988-1|

Attachments:

* [6988-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/16a/16a1747a6e18770c504e083b68e40ecfb62e3de7/51a13f679e74a8c76831dec21101a00b95088d0b.diff) (9159 bytes) - added by adiroiban on 2014-02-27 15:52:57Z - 
* [6988-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/16a/16a1747a6e18770c504e083b68e40ecfb62e3de7/34b8e096414f7016b10813c2a2c5fefebc08e318.diff) (9700 bytes) - added by adiroiban on 2014-11-10 15:56:14Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6988 6988
type__task task
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_python_pyflakes_6988_1 branches/python-pyflakes-6988-1
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1393501942000000 1393501942000000
changetime__1424126656363292 1424126656363292
version__None None
owner__adiroiban adiroiban
cc__wsanchez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

For doc test I had to define variables used in exampamples.

For Shortcut() `arguments` argument I had no idea if this is list or a string... could not find any test for that.

There are still some warnings in threadable.py and I don't know how to fix them:
```
twisted/python/threadable.py:26: undefined name 'XLock'
twisted/python/threadable.py:35: undefined name '_synchLockCreator'
twisted/python/threadable.py:37: undefined name 'XLock'
twisted/python/threadable.py:38: undefined name '_synchLockCreator'
twisted/python/threadable.py:139: undefined name 'XLock' in __all__
```

twisted/python/compat.py:95: undefined name 'bestLen' I had no idea what default value to use.

Hope that during review the reviewer will suggest how to fix the remaining errors / or fix them.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [43291]) Branching to python-pyflakes-6988.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @adiroiban|
|-|-|

Sorry for the latency on this.

Thanks again for your contributions to getting the code cleaned up.

It looks like the patch doesn't apply totally cleanly, and some of the cleanups break the tests, so they must not be just cleanups :) (as you can see on the py-without-modules buildbot).

I'll try to get around to re-reviewing more quickly in the future.  Feel free to ping me on IRC to remind me :-).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review.

I broke the threadpool... this is the first time than pyflakes is not detecting an used variable... but the code is also a bit strange

There are still some errors which I don't know how to fix as described in comment 2 [#6988](https://github.com/twisted/twisted/issues/6988)#comment:2

Maybe a separate ticket should be created for that... as the code looks broken.. so I assume that those lines are no longer used.

I have attached a new diff based on latest trunk.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_9"></a>|@wsanchez commented|
|-|-|

(In [43460]) New branch for 6988

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_10"></a>|@wsanchez commented|
|-|-|

This improves things, with some exceptions noted above:

```
[pucca:~/Developer/Twisted/python-pyflakes-6988-1] wsanchez$ pyflakes twisted/python
twisted/python/_shellcomp.py:294: local variable 'aCD' is assigned to but never used
twisted/python/_shellcomp.py:384: local variable 'descr' is assigned to but never used
twisted/python/compat.py:95: undefined name 'bestLen'
twisted/python/hook.py:130: local variable 'phaselistname' is assigned to but never used
twisted/python/threadable.py:26: undefined name 'XLock'
twisted/python/threadable.py:35: undefined name '_synchLockCreator'
twisted/python/threadable.py:37: undefined name 'XLock'
twisted/python/threadable.py:38: undefined name '_synchLockCreator'
twisted/python/threadable.py:139: undefined name 'XLock' in __all__
twisted/python/threadpool.py:217: local variable 'o' is assigned to but never used
twisted/python/test/test_components.py:146: local variable 'co2' is assigned to but never used
twisted/python/test/test_hashlib.py:20: 'hashlib' imported but unused
twisted/python/test/test_runtime.py:103: local variable 'result' is assigned to but never used
twisted/python/test/test_runtime.py:119: 'threading' imported but unused
twisted/python/test/test_util.py:369: local variable 'InsensitiveDict' is assigned to but never used
```
```
[pucca:~/Developer/Twisted/python-pyflakes-6988-1] wsanchez$ patch -p0 < /Volumes/data/Users/wsanchez/Downloads/6988-2.diff 
patching file twisted/python/_shellcomp.py
patching file twisted/python/hook.py
patching file twisted/python/log.py
patching file twisted/python/shortcut.py
patching file twisted/python/test/test_components.py
patching file twisted/python/test/test_hashlib.py
patching file twisted/python/test/test_runtime.py
patching file twisted/python/test/test_util.py
patching file twisted/python/threadpool.py
```
```
[pucca:~/Developer/Twisted/python-pyflakes-6988-1] wsanchez$ pyflakes twisted/python
twisted/python/compat.py:95: undefined name 'bestLen'
twisted/python/threadable.py:26: undefined name 'XLock'
twisted/python/threadable.py:35: undefined name '_synchLockCreator'
twisted/python/threadable.py:37: undefined name 'XLock'
twisted/python/threadable.py:38: undefined name '_synchLockCreator'
twisted/python/threadable.py:139: undefined name 'XLock' in __all__
```

You might want to open another ticket to follow up on the remaining bits.

I see the dilemma with `bestLen`; not the clearest code to follow.  It seems clear that the left side of that `or` is not expected to be reached before `bestLen` is set, or you'r get an exception.  Perhaps that means `None` is a reasonable default here, and the `>` operator should puke on that.  Same in line 99.  So I guess `None` seems OK there.

Looking at threadable.py… harm, yuck.  Looks like what's going on with is that `threaded`, `_synchLockCreator`, and `XLock` are all being set in the `init` function.  `threaded` is set to `None` at the bottom of the file.  Doing the same with the others seems to make pyflakes happy and tests still pass.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_11"></a>|@wsanchez commented|
|-|-|

r43461 merges in your patch to [branches/python-pyflakes-6988-1.](branches/python-pyflakes-6988-1.)

r43462 reverts the change to shortcut.py.  We don't think adding noise to docstrings to silence pyflakes is a good trade-off, and apparently this is now fixed in pyflakes.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez commented|
|-|-|

The rest looks good.  Kicking the builders.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez set owner to @adiroiban<br>@wsanchez set status to `new`|
|-|-|

These tests fail:

```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/python/test/test_hashlib.py", line 24, in test_deprecation
    self.assertIdentical(warnings[0]['category'], DeprecationWarning)
exceptions.IndexError: list index out of range

twisted.python.test.test_hashlib.HashObjectTests.test_deprecation
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/python/test/test_util.py", line 953, in test_addPluginDir
    self.assertEqual(len(warnings), 1)
exceptions.TypeError: object of type 'module' has no len()

twisted.python.test.test_util.DeprecationTests.test_addPluginDir
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/buildslave/run/debian6-x86_64-py2.6-glib/Twisted/twisted/python/test/test_util.py", line 370, in test_preserve
    self.assertEqual(eval(repr(dct)), dct)
  File "<string>", line 1, in <module>
    
exceptions.NameError: name 'InsensitiveDict' is not defined

twisted.python.test.test_util.InsensitiveDictTest.test_preserve
-------------------------------------------------------------------------------
```

So you'll need to address those.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban removed owner|
|-|-|

I have no idea why `namedModule` did not work and `__import__` did, in twisted.python.test.test_hashlib.HashObjectTests.test_deprecation as namedModule is just calling __import__

--------

For twisted.python.test.test_util.InsensitiveDictTest.test_preserve instead of using the default globals() I went for specific globals.

------

I have fixed the remaining errors as suggested by wsanchez

on my local computer I get clean results, my bad for not running those test before sending the patch


```
$ pyflakes twisted/python/
$ ./bin/trial twisted.python.test
```

I will also send my latest patch to buildbot.

This needs to be reviewed again and with great care.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

some builders that a long time to end... but most builder pass, so sending this for review. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

> I have no idea why namedModule did not work and __import__ did, in twisted.python.test.test_hashlib.HashObjectTests.test_deprecation as namedModule is just calling import

Probably because of stack frames.  The deprecation system works by walking up the call stack a specified number of frames.  This makes it very fragile with respect to how a deprecated API is called.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_17"></a>|@wsanchez set owner to @adiroiban|
|-|-|

I'm not fond of the `twisted/python/threadpool.py` change.  Doesn't that one seem like a bug in pyflakes?  May be best to leave that error be.  I don't think we like putting noise into code to please the linter.

There was a buildbot fail in `py_without_modules`, but a rebuild cleared it.

Super minor nit: in `twisted/python/threadable.py`, in your comment, "this" should be "these" and a comma should follow "globals".  Meh.

Rest looks good.  Nice cleanup to the use of `eval` in `test_util.py`, by the way.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review.

I assume you are not happy with the  changes in ```twisted/python/threadable.py```

I don't think this is an error in pylint.
Explicit is better than implicit, and so is better to first define them in the global scope, rather than define them via an if branch of a function.

Those globals are not always overwritten in the init() function... so we might have code in which they are defined as a class or not defined at all.

I have updated the comment.

I have not removed the code from ```twisted/python/threadable.py``` and asking it for a re-review.

I prefer those comments rather than having to be constantly reminded about those lines when I run pyflake on my computer.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_19"></a>|@wsanchez commented|
|-|-|

It's not a bug in pyflakes that it doesn't notice that `o` is used as a loop variable?

"Fake usage to fool pyflakes" sounds a lot like "work around a bug in pyflakes".

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_21"></a>|@adiroiban commented|
|-|-|

@wsanchez True. I have reverted that change.

Will try to have it fixed in upstream pyflakes.

I have also created the newfile and sent the branch to buildbot.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_22"></a>|@wsanchez set owner to @adiroiban|
|-|-|

Cool, reviewed.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_23"></a>|@adiroiban set status to `closed`|
|-|-|

(In [43933]) Merging python-pyflakes-6988-1: Clean pyflakes for twisted core.

Author: adiroiban
Reviewer: wsanchez, glyph
Fixes: [#6988](https://github.com/twisted/twisted/issues/6988)
